runtime/src/main/java/org/corfudb/protocols/wireprotocol/NettyCorfuMessageDecoder.java

```diff
@@ -1,25 +1,66 @@
 package org.corfudb.protocols.wireprotocol;
 
 import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufInputStream;
 import io.netty.buffer.Unpooled;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.handler.codec.ByteToMessageDecoder;
 
+import java.io.IOException;
 import java.util.List;
-
 import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.proto.service.CorfuMessage.RequestMsg;
+import org.corfudb.runtime.proto.service.CorfuMessage.ResponseMsg;
 
+import static org.corfudb.protocols.CorfuProtocolCommon.LEGACY_CORFU_MSG_MARK;
+import static org.corfudb.protocols.CorfuProtocolCommon.PROTO_CORFU_REQUEST_MSG_MARK;
+import static org.corfudb.protocols.CorfuProtocolCommon.PROTO_CORFU_RESPONSE_MSG_MARK;
 
 /**
  * Created by mwei on 10/1/15.
  */
 @Slf4j
 public class NettyCorfuMessageDecoder extends ByteToMessageDecoder {
 
+    /**
+     * Decodes an inbound corfu message from a ByteBuf. The corfu message is either
+     * legacy (of type CorfuMsg) or Protobuf (of type RequestMsg/ResponseMsg).
+     *
+     * @param channelHandlerContext   the Netty channel handler context
+     * @param byteBuf                 the underlying ByteBuf
+     * @param list                    a list of decoded objects given to the
+     *                                next pipeline handler
+     */
     @Override
     protected void decode(ChannelHandlerContext channelHandlerContext, ByteBuf byteBuf,
                           List<Object> list) throws Exception {
-        list.add(CorfuMsg.deserialize(byteBuf));
+        // Check the type of message based on first byte
+        byte msgMark = byteBuf.readByte();
+
+        if (msgMark == LEGACY_CORFU_MSG_MARK) {
```

Comment:
Would be great to have an enum here (I mentioned above) and replace `if` conditions with `switch` statement

Reply:
Done.